Source/SwiftLintFramework/Models/LinterCache.swift

```diff
@@ -65,31 +70,47 @@ public final class LinterCache {
 
         let configurationDescription = configuration.cacheDescription
 
-        lock.lock()
-
-        guard let filesCache = cache[configurationDescription] as? [String: Any],
+        guard let filesCache = readCache[configurationDescription] as? [String: Any],
             let entry = filesCache[file] as? [String: Any],
             let cacheLastModification = entry[Key.lastModification.rawValue] as? TimeInterval,
             cacheLastModification == lastModification.timeIntervalSinceReferenceDate,
             let violations = entry[Key.violations.rawValue] as? [[String: Any]] else {
-                lock.unlock()
                 return nil
         }
 
-        lock.unlock()
         return violations.flatMap { StyleViolation.from(cache: $0, file: file) }
     }
 
     public func save() throws {
         guard let url = location else {
             throw LinterCacheError.noLocation
         }
+        guard !writeCache.isEmpty else {
+            return
+        }
+        flushWriteCacheToReadCache()
         lock.lock()
```

Comment:
Actually, we can make `save()` thread-safe if we think about a way to support the unit tests. Maybe something like this

```swift
func flushWriteCacheToReadCache() {
    readCache = merge()
}

func merge() -> [String: Any] {
  // ...
}

func save() {
   // ...
    let cache = merge()
    let json = toJSON(cache)
   // ...
}
```

(please ignore the bad names and lack of details)